### PR TITLE
chore: add bn254 attribute when needed in the stdlib

### PR DESCRIPTION
**Pull request by guipublic**
# Description

Use the field attribute to restrict some functions in stdlib to bn254 field

## Problem\*

Some stdlib functions require that noir native field is the one from bn254 curve, so we use the new field attribute so that these functions cannot be used if the field is different.

## Summary\*

Add the attribute `#[field(bn254)]` to a few functions in stdlib, such as:
poseidon, babyjubjub and mimc.

## Documentation

- [ ] This PR requires documentation updates when merged.

  

  - [ ] I will submit a noir-lang/docs PR.

  

  - [ ] I will request for and support Dev Rel's help in documenting this PR.

  
  

## Additional Context

# PR Checklist\*

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: noir_stdlib/src/hash/poseidon/bn254/perm.nr [hunk 1]
```diff
@@ -3,6 +3,7 @@ use crate::hash::poseidon::bn254::consts;
 use crate::hash::poseidon::bn254::permute;
 use crate::hash::poseidon::PoseidonConfig;
 
+#[field(bn254)]
```
Comment (kevaundray) [inline noir_stdlib/src/hash/poseidon/bn254/perm.nr:6]: seems like it would be good if we could add this to the bn254 module instead?

**Review by jfecher — APPROVED** [review]
LGTM. Pinging @vezenovm and @kevaundray for re-review just in case.

**Comment by guipublic** [comment]
> I'm fine with merging this right now, but it does feel that the attributes in `noir_stdlib/src/hash/poseidon/bn254/perm.nr` are unneeded as they should be encapsulated by the bn254 module instead.

But we do not have attribute for modules?

**Comment by vezenovm** [comment]
> But we do not have attribute for modules?

Yeah we do not, but each method calls `permute` from the `bn254` module which has the attribute on it. It would be nice to be able to specif on the module itself as well, but I think this is good for now.

File: noir_stdlib/src/hash/poseidon/bn254/perm.nr (hunk 1 above)
Comment (guipublic) [reply noir_stdlib/src/hash/poseidon/bn254/perm.nr:6]: I am closing this because I have added an issue for having the field attribute on a module.

**Review by vezenovm — APPROVED** [review]

**Comment by kevaundray** [comment]
> > I'm fine with merging this right now, but it does feel that the attributes in `noir_stdlib/src/hash/poseidon/bn254/perm.nr` are unneeded as they should be encapsulated by the bn254 module instead.
> 
> But we do not have attribute for modules?

Yeah it would need to be introduced in a separate PR

**Pull request merged**